qualtran/surface_code/multi_factory.py

```diff
@@ -0,0 +1,41 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from qualtran.surface_code.algorithm_summary import AlgorithmSummary
+from qualtran.surface_code.magic_state_factory import MagicStateFactory
+
+from attrs import frozen
+
+
+@frozen
+class MultiFactory(MagicStateFactory):
+    """Overlay of MagicStateFactory representing multiple factories of the same kind.
+
+    Args:
+        base_factory (MagicStateFactory): the base factory to be replicated.
+        n_factories (int): number of factories to construct.
+    """
+
+    base_factory: MagicStateFactory
+    n_factories: int
+
+    def footprint(self) -> int:
+        return self.base_factory.footprint() * self.n_factories
+
+    def n_cycles(self, n_magic: AlgorithmSummary) -> int:
+        return self.base_factory.n_cycles(n_magic) // self.n_factories
```

Comment:
hmm, `//` does integer division which is similar to `floor`. I think technically we want `ceil`